qiskit/circuit/quantumcircuit.py

```diff
@@ -1756,22 +1756,29 @@ def draw(
             cregbundle=cregbundle,
         )
 
-    def size(self) -> int:
-        """Returns total number of gate operations in circuit.
+    def size(self, filter_function: Optional[callable] = lambda x: not x[0]._directive) -> int:
+        """Returns total number of instructions in circuit.
+
+        Args:
+            filter_function (callable): a function to filter out some instructions.
+                Should take as input a tuple of (Instruction, list(Qubit), list(Clbit)).
+                By default filters out "directives", such as barrier or snapshot.
 
         Returns:
             int: Total number of gate operations.
         """
-        gate_ops = 0
-        for instr, _, _ in self._data:
-            if not instr._directive:
-                gate_ops += 1
-        return gate_ops
+        ops = 0
+        for instr, _, _ in list(filter(filter_function, self._data)):
```

Comment:
No need to use a list here we can just iterate over the filter directly

```suggestion
        for instr, _, _ in filter(filter_function, self._data):
```